File: src/stages/query-generation/strategies/parse.ts
import { randomBytes } from 'node:crypto';
import type { GeneratedQuery, QueryCategory, QueryStrategy } from '../../../types/query.js';

const VALID_CATEGORIES = new Set<QueryCategory>([
  'recommendation', 'comparison', 'best-of', 'how-to', 'review', 'alternative', 'general',
]);

/**
 * Parse AI-generated query lines into GeneratedQuery objects.
 * Expected format per line: <query text> | <category> | <intent> [| <persona>]
 * Gracefully handles malformed lines by using defaults.
 */
export function parseAIQueries(
  text: string,
  strategy: QueryStrategy,
  prefix: string,
  maxCount: number,
): GeneratedQuery[] {
  const queries: GeneratedQuery[] = [];
  const lines = text.split('\n').map((l) => l.trim()).filter(Boolean);

  for (const line of lines) {
    if (queries.length >= maxCount) break;

    // Skip lines that look like headers, numbering prefixes, or markdown
    const cleaned = line.replace(/^\d+[\.\)]\s*/, '').replace(/^[-*]\s*/, '').trim();
    if (!cleaned || cleaned.startsWith('#') || cleaned.startsWith('```')) continue;

    const parts = cleaned.split('|').map((p) => p.trim());
    if (parts.length < 2) continue;

    const queryText = parts[0];
    if (!queryText || queryText.length < 10) continue; // skip very short/empty lines

    const rawCategory = parts[1]?.toLowerCase().trim() as QueryCategory;
    const category: QueryCategory = VALID_CATEGORIES.has(rawCategory) ? rawCategory : 'general';
    const intent = parts[2]?.toLowerCase().trim() || 'general';
    const persona = parts[3]?.toLowerCase().trim();

    const metadata: Record<string, unknown> = {};
    if (persona) metadata.persona = persona;

    queries.push({
      id: `${prefix}-${randomBytes(4).toString('hex')}`,
      text: queryText,
      category,
      strategy,
      intent,
      ...(Object.keys(metadata).length > 0 ? { metadata } : {}),
    });
  }

  return queries;
}

File: src/voyage_geo/stages/query_generation/strategies/parse.py
"""Parse AI-generated query text into GeneratedQuery objects."""

from __future__ import annotations

import secrets
from typing import get_args

from voyage_geo.types.query import GeneratedQuery, QueryCategory, QueryStrategy

VALID_CATEGORIES: set[str] = set(get_args(QueryCategory))


def parse_ai_queries(
    text: str,
    strategy: QueryStrategy,
    prefix: str,
    max_count: int,
) -> list[GeneratedQuery]:
    queries: list[GeneratedQuery] = []
    lines = [line.strip() for line in text.split("\n") if line.strip()]

    for line in lines:
        if len(queries) >= max_count:
            break

        # Strip numbering and bullets
        import re
        cleaned = re.sub(r"^\d+[.)]\s*", "", line)
        cleaned = re.sub(r"^[-*]\s*", "", cleaned).strip()
        if not cleaned or cleaned.startswith("#") or cleaned.startswith("```"):
            continue

        parts = [p.strip() for p in cleaned.split("|")]
        if len(parts) < 2:
            continue

        query_text = parts[0]
        if not query_text or len(query_text) < 10:
            continue

        raw_category = parts[1].lower().strip()
        category: QueryCategory = raw_category if raw_category in VALID_CATEGORIES else "general"  # type: ignore[assignment]
        intent = parts[2].lower().strip() if len(parts) > 2 else "general"
        persona = parts[3].lower().strip() if len(parts) > 3 else None

        metadata: dict = {}
        if persona:
            metadata["persona"] = persona

        queries.append(
            GeneratedQuery(
                id=f"{prefix}-{secrets.token_hex(4)}",
                text=query_text,
                category=category,
                strategy=strategy,
                intent=intent,
                metadata=metadata or None,
            )
        )

    return queries
